src/libsyntax/codemap.rs

```diff
@@ -576,6 +577,42 @@ impl CodeMap {
         self.span_until_char(sp, '{')
     }
 
+    /// Returns a new span representing just the end-point of this span
+    pub fn end_point(&self, sp: Span) -> Span {
+        let hi = sp.hi().0.checked_sub(1).unwrap_or(sp.hi().0);
+        let hi = self.get_start_of_char_bytepos(BytePos(hi));
+        let lo = cmp::max(hi.0, sp.lo().0);
+        sp.with_lo(BytePos(lo))
+    }
+
+    /// Returns a new span representing the next character after the end-point of this span
+    pub fn next_point(&self, sp: Span) -> Span {
+        let hi = sp.lo().0.checked_add(1).unwrap_or(sp.lo().0);
+        let hi = self.get_start_of_char_bytepos(BytePos(hi));
+        let lo = cmp::max(sp.hi().0, hi.0);
+        Span::new(BytePos(lo), BytePos(lo), sp.ctxt())
+    }
+
+    fn get_start_of_char_bytepos(&self, bpos: BytePos) -> BytePos {
+        let idx = self.lookup_filemap_idx(bpos);
+        let files = self.files.borrow();
+        let map = &(*files)[idx];
+
+        for mbc in map.multibyte_chars.borrow().iter() {
+            if mbc.pos < bpos {
```

Comment:
I think I'd probably be able to get the "subtract one and go further back if needed" strategy working, but when I implemented it, I made a logic error, it would compile but subsequent compiles would fall into an infinite loop. In a subsequent attempt, I took the approach currently in the PR.

I'd be happy to take another go at it if you'd like.

Reply:
Pushed the "subtract one and go further back if needed" version.